### PR TITLE
Update sampler.py

### PR DESCRIPTION
**Pull request by weifanfrank**
1. Import modules
2. Add code "Compute_core_ratio"

## Summary

Include a summary of major changes in bullet points:

- Feature 1
- Fix 1

## Additional dependencies introduced (if any)

- List all new dependencies needed and justify why. While adding dependencies that bring
  significantly useful functionality is perfectly fine, adding ones that add trivial
  functionality, e.g., to use one single easily implementable function, is frowned upon.
  Provide a justification why that dependency is needed. Especially frowned upon are
  circular dependencies.

## TODO (if any)

If this is a work-in-progress, write something about what else needs to be done.

- Feature 1 supports A, but not B.

### CODE REVIEW

**Comment by mjwen** [comment]
All the functions are empty, I just defined what they look like... you need to add the contents. You can do whatever you want there to make it work (you can test it by using the sampler).

File: src/potdata/sampler.py [hunk 1]
```diff
@@ -175,10 +182,71 @@ def __init__(
...
+            species=["Li", "Y", "Cl"],
+            periodic=True, # Set periodicity to True for a periodic system
+            r_cut=5.0,
+            n_max=8,
+            l_max=6,
+            sigma=0.1
+        )
+
+        # Read the trajectory file
+        traj = Trajectory(traj_file)
```
Comment (mjwen) [inline src/potdata/sampler.py:204]: where does `traj_file` come from?

File: src/potdata/sampler.py [hunk 2]
```diff
@@ -175,10 +182,71 @@ def __init__(
         self.reachable_ratio = reachable_ratio
         self.core_ratio = core_ratio
 
-        self._indices: list[int] = None
+        self._indices: list[int] = []
 
         np.random.seed(seed)
 
+    def _get_soap_vectors(self, data: list[Structure]) -> list[np.ndarray]:
```
Comment (mjwen) [inline src/potdata/sampler.py:189]: This function corresponds to step 1 above. 

- You should only create all the soap vectors for each `Structure`, not do any subselection, that is for step 2.
- The input is a list of pymatgen`Structure`. you can use `ASEAtomsAdaptor` to convert it to ASE Atoms. Then you can provide it to  `soap_desc.create`
- You should not do any hard code. stuff like `traj[-6000::5]` are absolute hard coding. This function should be general. We will use other sampler to select the steps. So, here you covert all the steps into soap vectors.

File: src/potdata/sampler.py [hunk 3]
```diff
@@ -175,10 +182,71 @@ def __init__(
...
+        """Convert structures to SOAP vectors.
+           Take Li atom in Li3YCl6 as an example."""
```
Comment (mjwen) [inline src/potdata/sampler.py:191]: Should not hard code it, for example, the species should be obtained from the structures.

File: src/potdata/sampler.py [hunk 4]
```diff
@@ -175,10 +182,71 @@ def __init__(
...
+
+        # Save the SOAP vectors to a file
+        np.save(save_file, soap_vectors_array)
+
+        # Load the SOAP vectors from the .npy file
+        soap_vectors = np.load(save_file)
+
+        return soap_vectors
+
+    def _dim_reduction(self, vectors: list[np.ndarray], dim: int):
```
Comment (mjwen) [inline src/potdata/sampler.py:228]: This corresponds to step 4 above. 

- You should use `dim` as `n_components`, not hard code it.

File: src/potdata/sampler.py [hunk 5]
```diff
@@ -175,10 +182,71 @@ def __init__(
...
+        """Perform DBSCAN."""
+        eps = self.db_kwargs.get('eps', 0.00019)
+        min_samples = self.db_kwargs.get('min_samples', 4)
+
+        db = DBSCAN(eps=eps, min_samples=min_samples).fit(data)
+        labels = db.labels_
+
+        return labels
+
     def sample(self, data: list[Structure]) -> list[Structure]:
```
Comment (mjwen) [inline src/potdata/sampler.py:250]: This is the overall function that you would finally call. This basically calls the above 6 steps. You need to finish this function. The logic is there, you just need to add a few stuff. 

In the end, you just need to call this function, and everything should be done. You need to test it. 

Line 255 is step 2 above. 

Line 270 needs to be filled.

File: src/potdata/sampler.py [hunk 6]
```diff
@@ -206,23 +274,134 @@ def sample(self, data: list[Structure]) -> list[Structure]:
...
-    def _dim_reduction(self, vectors: list[np.ndarray], dim: int):
-        """Perform dimension reduction on the SOAP vectors."""
+        vectors = np.concatenate(soap_vectors, axis=1)
 
-    def _cluster(self, data: list[Structure]):
-        """Perform DBSCAN."""
+        if self.pca_dim is not None:
+            vectors = self._dim_reduction(vectors, self.pca_dim)
+
+        db = DBSCAN(eps=eps, min_samples=min_samples).fit(vectors)
```
Comment (mjwen) [inline src/potdata/sampler.py:318]: You should not call any DBSCAN here. In the previous steps, you've already obtained the info such as the labels. Here you just plot them.

File: src/potdata/sampler.py [hunk 7]
```diff
@@ -206,23 +274,134 @@ def sample(self, data: list[Structure]) -> list[Structure]:
...
+    def _compute_core_ratio(self, soap_vectors):
         """Compute the ratio of core data points to sample.
 
         ratio = min_samples/average_num_neighbors
         """
+        
+        eps = self.db_kwargs.get('eps', 0.00019)
+        min_samples = self.db_kwargs.get('min_samples', 4)
+
+        db = DBSCAN(eps=eps, min_samples=min_samples).fit(soap_vectors)
```
Comment (mjwen) [inline src/potdata/sampler.py:387]: Again, you should not call DBSCAN here. it should already be done in step 5. You just need to use the results there.

File: src/potdata/sampler.py (hunk 2 above)
Comment (weifanfrank) [reply src/potdata/sampler.py:190]: 1. Fixed, separate the original codes to step 1 and step 2 individually
2. Fixed
3. Fixed

File: src/potdata/sampler.py (hunk 3 above)
Comment (weifanfrank) [reply src/potdata/sampler.py]: Fixed

File: src/potdata/sampler.py (hunk 4 above)
Comment (weifanfrank) [reply src/potdata/sampler.py:247]: Fixed

File: src/potdata/sampler.py (hunk 6 above)
Comment (weifanfrank) [reply src/potdata/sampler.py]: Fixed

File: src/potdata/sampler.py (hunk 7 above)
Comment (weifanfrank) [reply src/potdata/sampler.py]: Fixed

**Pull request closed without merging**